M2/Macaulay2/packages/HomologicalShiftIdeals.m2

```diff
@@ -0,0 +1,949 @@
+    );
+
+-------------------------------------------------------------------------------------------
+-- Whether a monomial ideal and its homological shift ideals have linear quotients
+-------------------------------------------------------------------------------------------
+mydoc = concatenate(mydoc,///
+Node
+  Key
+    hasHomologicalLinearQuotients
+    (hasHomologicalLinearQuotients,Ideal)
+  Headline
+    checks if a monomial ideal has homological linear quotients
+  Usage
+    hasLinearQuotients(I)
+  Inputs
+    I:Ideal
+      a monomial ideal $I\subset S=K[x_1,\dots,x_n]$, $K$ a field
+  Outputs
+    b:Boolean
+      whether $I$ has homological linear quotients
+  Description
+    Text
+      A monomial ideal $I$ of a polynomial ring $S=K[x_1,\ldots,x_n]$ has homological linear quotients if $\text{HS}_k(I)$ has linear quotients for all $k$.
+    Example
+      S = QQ[x_1..x_5];
+      I = ideal(x_1,x_2,x_3,x_4)*ideal(x_3,x_4,x_5);
+      hasHomologicalLinearQuotients I
+    Example
+      S = QQ[x,y];
+      I = ideal(x^2,y^2);
+      hasHomologicalLinearQuotients I
+///);
+
+hasHomologicalLinearQuotients = method(TypicalValue=>Boolean)
+hasHomologicalLinearQuotients (Ideal) := I -> (
+        if not isMonomialIdeal(I) then error "expected a monomial ideal";
+        S:=ring I;
+        p:=pdim module I;
+        i:=0;
+        while i<p and hasLinearQuotients(HS(I,i)) do (
+             i=i+1;
+        );
+        hasLinearQuotients(HS(I,i))
+    );
+
+-------------------------------------------------------------------------------------------
+--  Compute an admissible order of a monomial ideal with linear quotients
+-------------------------------------------------------------------------------------------
+mydoc = concatenate(mydoc,///
+Node
+  Key
+    admissibleOrder
+    (admissibleOrder,Ideal)
+  Headline
+    gives an admissible order of a monomial ideal having linear quotients
+  Usage
+    admissibleOrder(I)
+  Inputs
+    I:Ideal
+      a monomial ideal $I\subset S=K[x_1,\dots,x_n]$, $K$ a field
+  Outputs
+    l:List
+      an admissible order
+  Description
+    Text
+      Let $I$ be a monomial ideal of a polynomial ring $S=K[x_1,\ldots,x_n]$ having linear quotients. Any linear quotients order $u_1,\dots,u_m$ is called an admissible order. This method determines if a given order of $G(I)$ is an admissible order of $I$.
+    Example
+      S = QQ[x_1..x_5];
+      I = ideal(x_1,x_2,x_3,x_4)*ideal(x_3,x_4,x_5);
+      admissibleOrder I
+    Example
+      S = QQ[x,y];
+      I = ideal(x^2,y^2);
+      admissibleOrder I
+///);
+
+admissibleOrder = method(TypicalValue=>List)
+admissibleOrder (Ideal) := I -> (
+        if not isMonomialIdeal(I) then error "expected a monomial ideal";
+        if not hasLinearQuotients(I) then error "the ideal does not have linear quotients";
+        S:=ring I;
+        n:=numgens S-1;
+        bDeg:=boundingMultidegree I;
+        L:=for i to n list for j to bDeg#i-1 list S_i;
```

Comment:
Updating this to `for ... list` broke `admissibleOrder` since now `L` is a nested list.  You could call `flatten` here to restore the old behavior.